tests/test_permissions.py

```diff
@@ -396,7 +400,9 @@ def test_cannot_read_list_permissions(self):
         self.assertListEqual(response.data, [])
 
     def test_cannot_method_not_allowed(self):
-        request = factory.generic('METHOD_NOT_ALLOWED', '/')
+        request = factory.generic(
+            'METHOD_NOT_ALLOWED', '/', HTTP_AUTHORIZATION=self.credentials['readonly']
+        )
```

Comment:
Are the test changes necessary? The changes to `DjangoModelPermissions` seems like it wouldn't affect existing behavior.

Reply:
Unfortunately, yes. It was a surprise for me, too.

This tests became broken, since they used an anonymous users (accidentally, I think) and started to retrieve "403 forbidden" instead of "405 not allowed". That's connected with the priority of permission checks, which is higher than most of other checks.

So, all I did - just changed users to authenticated ones in these tests.

Reply:
Will need to review it, but thanks for the explanation.

Reply:
So, it is possible to retain the existing behavior by moving the HTTP method check out of `get_required_permissions()` and into `has_permission()` before the `request.user` check. 

That said, I think the test changes are appropriate, as I'd argue that users shouldn't really be relying on the permissions class to perform 405 checks. 405 checking should happen earlier during dispatch.